store/copr/batch_coprocessor.go

```diff
@@ -95,9 +97,133 @@ func (rs *batchCopResponse) RespTime() time.Duration {
+			} else {
+				// if more than one store is valid, put the region
+				// to store candidate map
+				totalRegionCandidateNum += validStoreNum
+				totalRemainingRegionNum += 1
+				taskKey := ri.Region.String()
+				for _, storeID := range ri.AllStores {
+					if candidateMap, ok := storeCandidateRegionMap[storeID]; ok {
+						if _, ok := candidateMap[taskKey]; ok {
+							// duplicated region, should not happen, just give up balance
+							return originalTasks
+						}
+						candidateMap[taskKey] = ri
+					}
+				}
+			}
+		}
+	}
+	if totalRemainingRegionNum == 0 {
+		return originalTasks
+	}
+
+	avgStorePerRegion := float64(totalRegionCandidateNum) / float64(totalRemainingRegionNum)
+	findNextStore := func() uint64 {
+		store := uint64(math.MaxUint64)
+		weightedRegionNum := float64(0)
+		for storeID := range storeTaskMap {
+			if store == uint64(math.MaxUint64) && len(storeCandidateRegionMap[storeID]) > 0 {
+				store = storeID
+				weightedRegionNum = float64(len(storeCandidateRegionMap[storeID]))/avgStorePerRegion + float64(len(storeTaskMap[storeID].regionInfos))
+			} else {
+				num := float64(len(storeCandidateRegionMap[storeID])) / avgStorePerRegion
+				if num == 0 {
+					continue
+				}
+				num += float64(len(storeTaskMap[storeID].regionInfos))
+				if num < weightedRegionNum {
+					store = storeID
+					weightedRegionNum = num
+				}
+			}
+		}
+		return store
+	}
+	store := findNextStore()
+	for totalRemainingRegionNum > 0 {
+		if len(storeCandidateRegionMap[store]) == 0 {
+			store = findNextStore()
+		}
+		if store == uint64(math.MaxUint64) {
+			break
+		}
+		for key, ri := range storeCandidateRegionMap[store] {
+			storeTaskMap[store].regionInfos = append(storeTaskMap[store].regionInfos, ri)
+			totalRemainingRegionNum--
+			for _, id := range ri.AllStores {
+				if _, ok := storeCandidateRegionMap[id]; ok {
+					delete(storeCandidateRegionMap[id], key)
+					totalRegionCandidateNum--
+				}
+			}
+			if totalRemainingRegionNum > 0 {
+				weightedTaskNum := float64(len(storeCandidateRegionMap[store]))/avgStorePerRegion + float64(len(storeTaskMap[store].regionInfos))
+				avgStorePerRegion = float64(totalRegionCandidateNum) / float64(totalRemainingRegionNum)
+				for _, id := range ri.AllStores {
+					// it is not optimal because we only check the stores that affected by this region, in fact in order
+					// to find out the store with the lowest weightedTaskNum, all stores should be checked, but I think
+					// check only the affected stores is more simple and will get a good enough result
+					if id != store && len(storeCandidateRegionMap[id]) > 0 && float64(len(storeCandidateRegionMap[id]))/avgStorePerRegion+float64(len(storeTaskMap[id].regionInfos)) <= weightedTaskNum {
+						store = id
+						weightedTaskNum = float64(len(storeCandidateRegionMap[id]))/avgStorePerRegion + float64(len(storeTaskMap[id].regionInfos))
+					}
+				}
+			}
+			break
+		}
+	}
```

Comment:
refactor these code to make it clearer. It is a typical problem, may be a better algo to solve it.

Reply:
In my opinion, the optimistic algo is equal to a match problem or max flow problem, of which time complexity is expensive. So the greedy method is ok with me.